Sparse MPI_Alltoallv for ParallelTransfer

iraikov opened the pull request:
Hello,

This is a pull request for an alternative of mpi_transfer that uses an MPI_Alltoallv implementation based on Isend/Waitall. The alternative mpi_transfer implementation can be enabled at runtime via a hoc flag, `nrn_sparse_partrans'. This implementation has proven necessary for our large-scale model of the dentate gyrus, which was experiencing significant performance issues with regular Alltoallv on Blue Waters and on Frontera with the intel 19 MPI stack. This "sparse" parallel transfer / alltoallv algorithm is suitable for large network models with small number of gap junctions that are spread across all ranks. The pull request also includes a python test script that implements a simple network with gap junctions. Thanks for the consideration of including this functionality in NEURON.

pramodk commented:
Thank you @iraikov for this PR! 

I completely agree with you about the scalability aspects of `AlltoAll/AlltoAllv`. We are also deploying next hpc system and one of the major performance issue for MPI benchmarks is these two MPI functions (and we use those in NEURON).

I think [the CI](https://travis-ci.com/github/neuronsimulator/nrn/jobs/357603948) is failing because some minor issue:

```
[ 99%] Linking CXX executable ../../bin/nrniv
../../lib/libnrniv.so: undefined reference to `nrnmpi_int_alltoallv_sparse'
../../lib/libnrniv.so: undefined reference to `nrnmpi_dbl_alltoallv_sparse'
```
(without looking into details, I think this appears when you use MPI dynamic build time option. e.g. with CMake build it's `-DNRN_ENABLE_MPI_DYNAMIC=ON`). Check if you can reproduce otherwise I can also help.

iraikov commented:
Hi @pramodk 
Thanks a lot for your comments! I am getting cmake errors when trying to build on Frontera with Intel MPI using any of NRN_ENABLE_MPI_DYNAMIC=ON or NRN_ENABLE_MPI_DYNAMIC=$MPIHOME or NRN_ENABLE_MPI_DYNAMIC=$MPIHOME/bin. It seems that the MPI dynamic configuration logic is not able to detect the libraries and include files for some reason, so the corresponding list variables are empty. Has it been tested with Intel MPI?
I am including my cmake command below for reference. If it is easier for troubleshooting, we can get you an account on Frontera.

`cmake ../nrn -DNRN_ENABLE_INTERVIEWS=OFF -DNRN_ENABLE_MPI_DYNAMIC=/opt/intel/compilers_and_libraries_2018.6.288/linux/mpi/intel64 -DNRN_ENABLE_MPI=ON -DNRN_ENABLE_RX3D=ON -DNRN_ENABLE_CORENEURON=ON -DNRN_ENABLE_PYTHON=ON -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DPYTHON_EXECUTABLE:FILEPATH=`which python3` -DCMAKE_INSTALL_PREFIX=$HOME/bin/nrnpython3`

pramodk commented:
@iraikov : I think you have to change cmake options a bit:

```
-DNRN_ENABLE_MPI_DYNAMIC=ON
-DNRN_MPI_DYNAMIC=/path-of-intel-mpi-library/include
```

Here is what I have on our system:

```
cmake .. -DNRN_ENABLE_INTERVIEWS=OFF -DNRN_ENABLE_MPI_DYNAMIC=ON -DNRN_MPI_DYNAMIC=/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/impi/2018.1.163/intel64/include -DNRN_ENABLE_MPI=ON -DNRN_ENABLE_RX3D=ON -DNRN_ENABLE_CORENEURON=ON -DNRN_ENABLE_PYTHON=ON -DCMAKE_C_COMPILER=icc -DCMAKE_CXX_COMPILER=icpc -DPYTHON_EXECUTABLE:FILEPATH=`which python3` -DCMAKE_INSTALL_PREFIX=`pwd`/install
```

and that gives:

```
-- The C compiler identification is Intel 19.0.4.20190416
-- The CXX compiler identification is Intel 19.0.4.20190416
-- Check for working C compiler: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/intel-19.0.4-3ahwr77o2x/bin/icc
-- Check for working C compiler: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/intel-19.0.4-3ahwr77o2x/bin/icc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/intel-19.0.4-3ahwr77o2x/bin/icpc
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/intel-19.0.4-3ahwr77o2x/bin/icpc -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Found BISON: /usr/bin/bison (found version "3.0.4")
-- Found FLEX: /usr/bin/flex (found version "2.5.37")
-- Found Readline: /usr/include
-- Found PythonInterp: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/bin/python3 (found version "3.7.4")
-- Found Cython: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/py-cython-0.29.13-trqyrkvozc/bin/cython (found version "0.29.13")
-- Found MPI_C: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/compilers_and_libraries_2018.1.163/linux/mpi/intel64/lib/release_mt/libmpi.so (found version "3.1")
-- Found MPI_CXX: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/compilers_and_libraries_2018.1.163/linux/mpi/intel64/lib/libmpicxx.so (found version "3.1")
-- Found MPI: TRUE (found version "3.1")
-- Found PythonInterp: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/bin/python3 (found suitable version "3.7.4", minimum required is "3")
-- Found PythonLibs: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib/libpython3.7m.so
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Building CoreNEURON from submodule
-- Sub-project : using coreneuron from from external/coreneuron
-- Found Random123: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron/external/Random123
-- Found CLI11: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron/external/CLI11
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.5") found components:  doxygen missing components:  dot
-- Found PythonInterp: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/bin/python3 (found version "3.7.4")
-- Found MOD2C: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron/external/mod2c
-- Using mod2c submodule from /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron/external/mod2c
-- CXX Compile flags from BUILD_TYPE (RELWITHDEBINFO): -g  -O2
--
-- Configured CoreNEURON 0.18.0
--
-- Some things you can do now:
-- --------------------+--------------------------------------------------------
-- Command             |   Description
-- --------------------+--------------------------------------------------------
-- make install        | Will install CoreNEURON to: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/build_intel/install
-- --------------------+--------------------------------------------------------
--  Build option       | Status
-- --------------------+--------------------------------------------------------
-- COMPILE FLAGS       | -g  -O2  -Wno-unknown-pragmas -qopenmp
-- MPI                 | ON
--   INC               | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/compilers_and_libraries_2018.1.163/linux/mpi/intel64/include
-- OpenMP              | ON
-- NMODL               | OFF
-- MOD2C               | /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron/external/mod2c
-- GPU Support         | OFF
-- Auto Timeout        | ON
-- Wrap exp()          | OFF
-- SplayTree Queue     | ON
-- NetReceive Buffer   | ON
-- Caliper             | OFF
-- Likwid              | OFF
-- Unit Tests          | OFF
-- Reporting           | OFF
-- --------------+--------------------------------------------------------------
--  See documentation : https://github.com/BlueBrain/CoreNeuron/
-- --------------+--------------------------------------------------------------
--
--
-- Configured NEURON 8.0
--
-- Some things you can do now:
-- --------------+--------------------------------------------------------------
-- Command       |   Description
-- --------------+--------------------------------------------------------------
-- make install  | Will install NEURON to: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/build_intel/install
--               | Change the install location of NEURON using:
--               |     cmake <src_path> -DCMAKE_INSTALL_PREFIX=<install_path>
-- make docs     | Build the API documentation
-- make uninstall| Removes files installed by make install (todo)
-- --------------+--------------------------------------------------------------
--  Build option | Status
-- --------------+--------------------------------------------------------------
-- BUILD_TYPE    | RelWithDebInfo (allowed: Custom;Debug;Release;RelWithDebInfo;Fast)
-- COMPILE FLAGS | -g  -O2
-- Shared        | ON
-- Binary special| OFF
-- Legacy FR     | ON
-- MPI           | ON
--   INC         | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/compilers_and_libraries_2018.1.163/linux/mpi/intel64/include
--   LIB         | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/compilers_and_libraries_2018.1.163/linux/mpi/intel64/lib/libmpicxx.so
--   DYNAMIC     | ON
--     LIBNAME   | nrnmpi_mpich
--     INC       | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/intel-mpi-2018.1.163-q6l6bn6igv/impi/2018.1.163/intel64/include
-- Python        | ON
--   EXE         | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/bin/python3
--   INC         | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/include/python3.7m
--   LIB         | /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib/libpython3.7m.so
--   MODULE      | ON
--   DYNAMIC     | OFF
-- Readline      | /usr/lib64/libreadline.so
-- Curses        | /usr/lib64/libcurses.so;/usr/lib64/libform.so
-- RX3D          | ON
--   OptLevel    | 0
-- Interviews    | OFF
-- CoreNEURON    | ON
--   PATH        | /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/external/coreneuron
-- --------------+--------------------------------------------------------------
--  See documentation : https://www.neuron.yale.edu/neuron/
-- --------------+--------------------------------------------------------------
--
-- Configuring done
-- Generating done
-- Build files have been written to: /gpfs/bbp.cscs.ch/project/proj16/kumbhar/pramod_scratch/nrn/build_intel
```

@nrnhines could provide some hints about what needs to be changed for dynamic mpi build if we add new MPI function like `nrnmpi_int_alltoallv_sparse`. (I think we need to add function declaration in `src/nrnmpi/nrnmpidec.h`).

nrnhines commented:
>what needs to be changed for dynamic mpi build if we add new MPI function

Normally:
Add the declarations to nrnmpidec.h
Implement in mpispike.c
Decide if the trivial case (one rank, no mpi available) still needs the function and, if so, implement trivially in nrnmpi_dynam_stub.c following the naming pattern there.

With the following change, I can build from:
```
cmake .. -DCMAKE_INSTALL_PREFIX=install -DNRN_ENABLE_MPI_DYNAMIC=ON
```
[partrans_alltoallv_sparse_patch.txt](https://github.com/neuronsimulator/nrn/files/4874932/partrans_alltoallv_sparse_patch.txt)

nrnhines commented on src/nrniv/have2want.cpp:
```diff
@@ -169,7 +169,7 @@ sprintf(buf, "key = %lld is wanted but does not exist\n", (long long)key);
   // with that key is coming from
   int* want_s_ownerranks = new int[want_s_displ[nhost]];
-  nrnmpi_int_alltoallv(want_r_ownerranks, want_r_cnt, want_r_displ,
+  nrnmpi_int_alltoallv_sparse(want_r_ownerranks, want_r_cnt, want_r_displ,
```
here is a place where nrn_sparse_partrans is needed.

nrnhines commented on src/nrniv/partrans.cpp:
```diff
@@ -27,10 +27,10 @@
 #else
 #define sgid_t int
-#define sgid_alltoallv nrnmpi_int_alltoallv
+#define sgid_alltoallv nrnmpi_int_alltoallv_sparse
```
I guess we have to create a couple of wrapper functions that use nrn_sparse_partrans.

iraikov commented:
Hi @pramodk @nrnhines 
Thanks a lot for your help and comments, I was able to build with the dynamic MPI option and indeed adding the sparse alltoall declarations to nrnmpidec.h allowed the compilation to go through. Regarding sgid_alltoallv, do I understand correctly that you would like to have a wrapper function that selects the alltoallv implementation based on the flag variable? I will add that as well. Are there some additional test scripts for ParallelTransfer that you would suggest running?

nrnhines commented:
> like to have a wrapper function that selects the alltoallv implementation based on the flag variable?

That is correct.
>some additional test scripts for ParallelTransfer

I have some but let's defer for now.

With respect to my suggestion about using nrn_sparse_partrans in the have2want.cpp, I think that is a good  beginning for now as otherwise the sparse use is turned on permanently. However, that is also used to setup the multisend spike exchange patterns which often are not sparse til one gets to the 100k rank domain. So it may be that ultimately we develop a different property setting style for have2want.cpp

I have to say that I'm surprised that vendor implementations of MPI haven't already taken possible sparseness of  alltoallv
into account, especially on HPC machines.  I guess I got spoiled by the Blue Gene series. Thanks for this pull request!  I think it would be useful for me to add an optional sparse arg to ParallelContext.alltoall and .py_alltoall

iraikov commented:
Hi @nrnhines 

Thanks, I have added a wrapper for for sgid_alltoallv -- it is currently in partrans.cpp, but let me know if a different location is more appropriate. 
> 
> > some additional test scripts for ParallelTransfer
> 
> I have some but let's defer for now.

Ok, just let me know when the time is right.

> 
> With respect to my suggestion about using nrn_sparse_partrans in the have2want.cpp, I think that is a good beginning for now as otherwise the sparse use is turned on permanently. However, that is also used to setup the multisend spike exchange patterns which often are not sparse til one gets to the 100k rank domain. So it may be that ultimately we develop a different property setting style for have2want.cpp
> 
Yes, I have added the necessary conditional in have2want.cpp.


> I have to say that I'm surprised that vendor implementations of MPI haven't already taken possible sparseness of alltoallv
> into account, especially on HPC machines. I guess I got spoiled by the Blue Gene series. Thanks for this pull request! I think it would be useful for me to add an optional sparse arg to ParallelContext.alltoall and .py_alltoall

I was not able to replicate this issue on Blue Waters outside of the psolve simulation loop. Standalone calls to Alltoallv with sparse or dense data worked fine, but with psolve and gap junctions there was a very significant performance degradation. If you remember, you helped me verify that the simulation was load balanced. Perhaps the Cray MPI stack makes algorithm selection decisions based on the MPI communication patterns and there was something in our model that triggered the wrong heuristic. On Frontera, there seems to be some general problem with Intel MPI and the network drivers for their interconnect, as many MPI collectives that involve tens of thousands of ranks exhibit serious performance issues. As I recall, Allgather/Allgatherv required selection of Intel's "topology-aware" algorithms, and I wasn't able to find a suitable configuration for Alltoallv that worked at full-scale. 

Of course, it is possible these were issues caused by the specific versions or configurations of the software stacks I was using. But as troubleshooting the vendor-specific configurations is a rather time-consuming process, I found it easier to switch to less-demanding MPI communication patterns.

pramodk commented:
It's true that BlueGenes had very fast collective operations (due to dedicated network path). But all AlltoAll/AlltoAllv are very demanding on large number of systems (with most MPI libraries). For example, here is one of the benchmarking results showing latency of `allgatherv` vs `alltoallv` on ~28k cores:

```
# OSU MPI Allgatherv Latency Test v5.6.1
# Size       Avg Latency(us)   Min Latency(us)   Max Latency(us)  Iterations
4                    3643.60           3403.80           3841.87          50
8                    3640.80           3348.46           3903.50          50
16                   3683.92           3519.26           3935.49          50
32                   4527.73           4193.96           4914.01          50
64                   5400.28           5180.58           5723.04          50
128                  7159.58           6842.74           7534.28          50
256                  9186.50           8395.11           9724.44          50
512                 14077.27          13748.92          14410.59          50
1024                19764.02          19102.69          20781.76          50
2048                38826.97          37358.73          40918.45          50
4096                78538.85          76742.43          81844.08          50

# OSU MPI All-to-Allv Personalized Exchange Latency Test v5.6.1
# Size       Avg Latency(us)   Min Latency(us)   Max Latency(us)  Iterations
4                  407741.03         398952.83         415571.78          50
8                  407921.73         399079.03         413950.33          50
16                 408089.51         399379.40         416211.94          50
32                 408335.60         399649.33         415101.48          50
64                 408201.53         399708.44         416420.49          50
128                398307.27         386459.04         404329.53          50
256                405191.97         389977.75         424001.24          50
512                558657.50         534460.22         581558.16          50
1024               893286.37         872199.83         915550.68          50
2048              1651195.56        1617284.81        1690219.03          50
4096              3470605.65        3416773.00        3505006.81          50
```

pramodk commented:
I am bit confused by travis build is failing with following error:

```
OK
6070+neurondemo -c 'demo(4)' -c 'run()' -c 'quit()'
6071Creating x86_64 directory for .o files.
6072
6073/home/travis/.local/lib/python3.6/site-packages/neuron/.data/share/nrn/demo/release
6074cabpump.mod cachan1.mod camchan.mod capump.mod invlfire.mod khhchan.mod mcna.mod nacaex.mod nachan.mod release.mod
6075cabpump.mod cachan1.mod camchan.mod capump.mod invlfire.mod khhchan.mod mcna.mod nacaex.mod nachan.mod release.mod
6076mkdir -p x86_64
6077 -> Compiling mod_func.c
6078gcc-8 -O2   -I. -I..   -I/home/travis/.local/lib/python3.6/site-packages/neuron/.data/include  -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent/include -I/usr/lib/x86_64-linux-gnu/openmpi/include -fPIC -c mod_func.c -o mod_func.o
6079 -> NMODL cabpump.mod
6080MODLUNIT=/home/travis/.local/lib/python3.6/site-packages/neuron/.data/share/nrn/lib/nrnunits.lib \
6081  /home/travis/.local/lib/python3.6/site-packages/neuron/.data/bin/nocmodl cabpump.mod
6082/home/travis/.local/lib/python3.6/site-packages/neuron/.data/bin/nrnmech_makefile:93: recipe for target 'x86_64/cachan1.c' failed
6083make: *** [x86_64/cachan1.c] Error 1
6084make: *** Waiting for unfinished jobs....
6085 -> NMODL camchan.mod
6086MODLUNIT=/home/travis/.local/lib/python3.6/site-packages/neuron/.data/share/nrn/lib/nrnunits.lib \
6087  /home/travis/.local/lib/python3.6/site-packages/neuron/.data/bin/nocmodl camchan.mod
6088Translating camchan.mod into camchan.c
6089Thread Safe
```

@nrnhines : any hint about this?

nrnhines commented:
Here is another patch for the travis failure of the autotools build using ```--without-paranrn```. As I'm not yet familiar with the best practices for offering changes to an external pull request I'm afraid I have to give you another patch to add to your repository version. 
[without-paranrn-patch.txt](https://github.com/neuronsimulator/nrn/files/4878603/without-paranrn-patch.txt)
Still trying to figure out the travis failure involving cachan1.mod

iraikov commented:
> Here is another patch for the travis failure of the autotools build using `--without-paranrn`. As I'm not yet familiar with the best practices for offering changes to an external pull request I'm afraid I have to give you another patch to add to your repository version.
> [without-paranrn-patch.txt](https://github.com/neuronsimulator/nrn/files/4878603/without-paranrn-patch.txt)
> Still trying to figure out the travis failure involving cachan1.mod

Hi @nrnhines really appreciate the fast help and nor problem about the patch. I suspect the best practice would be for you to duplicate the pull request branch and for me to update from it, but that would be about the same amount of work.

iraikov commented:
> It's true that BlueGenes had very fast collective operations (due to dedicated network path). But all AlltoAll/AlltoAllv are very demanding on large number of systems (with most MPI libraries). For example, here is one of the benchmarking results showing latency of `allgatherv` vs `alltoallv` on ~28k cores:

These are pretty dramatic numbers, but they are actually consistent with what I observed on Blue Waters (two orders of magnitude difference with 32768 cores).  Does the functionality currently exist to instrument NEURON's MPI calls so that it is easy to measure psolve's MPI performance?

nrnhines commented:
Now only one travis test is failing and that is for a wheel test. The actual error is at present very puzzling and unexpected. Before we investigate further I think it will be best if we eliminate the  "This branch is out-of-date with the base branch" and you can do that with ```git merge master``` and pushing the result. Then it will be easier to do a diff with respect to the last successful travis build.

nrnhines commented:
@pramodk The travis build failure moved from cachan1 to invlfire for 43ca3bb. So I think the failure messages don't really give a hint about the true issue.

pramodk commented:
@nrnhines : after merge with master, I think we are good. The [last CI failure](https://travis-ci.com/github/neuronsimulator/nrn/jobs/358036334) is:

```
Notice: Assignment to the GLOBAL variable, "vol", is not thread safe
6087Translating camchan.mod into camchan.c
6088Thread Safe
6089make: write error: stdout
```
and I think it was temporary glitch. I have restarted the build, let's see.

pramodk commented:
I am surprised by the fact that master is building without glitch but this PR is failing. I am having a look...

pramodk commented:
@iraikov @nrnhines : I created #631 by simply pushing this PR as separate branch under nrn repo and then created PR. I have no idea why this changes travis behaviour! 

If CI for #631 passes then we can close this PR and merge #631.  (we should copy description from this PR)
